lib/logstash/outputs/datadog_logs.rb

```diff
@@ -6,67 +6,193 @@
 
+  # Respect limit documented at https://docs.datadoghq.com/agent/logs/?tab=tailexistingfiles#send-logs-over-https
+  DD_MAX_BATCH_SIZE = 200
+
   config_name "datadog_logs"
 
   default :codec, "json"
 
   # Datadog configuration parameters
-  config :api_key,     :validate => :string,  :required => true
-  config :host,        :validate => :string,  :required => true, :default  => 'intake.logs.datadoghq.com'
-  config :port,        :validate => :number,  :required => true, :default  => 10516
-  config :use_ssl,     :validate => :boolean, :required => true, :default  => true
-  config :max_backoff, :validate => :number,  :required => true, :default  => 30
-  config :max_retries, :validate => :number,  :required => true, :default  => 5
+  config :api_key, :validate => :string, :required => true
+  config :host, :validate => :string, :required => true, :default => "http-intake.logs.datadoghq.com"
+  config :port, :validate => :number, :required => true, :default => 443
+  config :use_ssl, :validate => :boolean, :required => true, :default => true
+  config :max_backoff, :validate => :number, :required => true, :default => 30
+  config :max_retries, :validate => :number, :required => true, :default => 5
+  config :use_http, :validate => :boolean, :required => false, :default => true
+  config :use_compression, :validate => :boolean, :required => false, :default => true
+  config :compression_level, :validate => :number, :required => false, :default => 6
+  config :no_ssl_validation, :validate => :boolean, :required => false, :default => false
 
+  # Register the plugin to logstash
   public
   def register
-    require "socket"
-    client = nil
-    @codec.on_event do |event, payload|
-      message = "#{@api_key} #{payload}\n"
-      retries = 0
+    client ||= new_client(@logger, @api_key, @use_http, @use_ssl, @no_ssl_validation, @host, @port, @use_compression)
+    @codec.on_event do |_, payload|
+      payload = encode(payload, @use_http, @api_key)
+      if @use_compression and @use_http
+        payload = gzip_compress(payload, @compression_level)
+      end
+      client.send_retries(payload, @max_retries, @max_backoff)
+    end
+  end
+
+  # Process a set of log events
+  public
+  def multi_receive(events)
+    return if events.empty?
+    if @use_http
+      batches = batch_events(events, DD_MAX_BATCH_SIZE)
+      batches.each do |batched_event|
+        @codec.encode(batched_event)
+      end
+    else
+      events.each do |event|
+        @codec.encode(event)
+      end
+    end
+  end
+
+  # Encode payload for Datadog to the right format (no-op for HTTP)
+  def encode(payload, use_http, api_key)
+    if not use_http
+      return "#{api_key} #{payload}"
+    else
+      return payload
+    end
+  end
+
+  # Group events in batches
+  def batch_events(events, max_batch_size)
+    batches = []
+    current_batch = []
+    events.each_with_index do |event, i|
+      if i > 0 and i % max_batch_size == 0
```

Comment:
We should also check the max payload size in bytes, isn't it?

Reply:
Correct, updated. I've reworked a bit the code to ensure we serialize events before computing the size.